File: Python3/1001 - 1200/1005 Maximize Sum Of Array After K Negations.py
# 52ms, 94.38%; 13.9MB, 11.11%
class Solution:
    def largestSumAfterKNegations(self, A: list, K: int) -> int:
        res, neg0 = sum(A), [x for x in A if x <= 0]
        res -= 2 * sum(sorted(neg0)[:K])
        x = K - len(neg0)
        if x > 0 and x & 1: res -= 2 * abs(min(A, key=abs))
        return res
    

File: 1001 - 1200/1005. Maximize Sum Of Array After K Negations.cc
// 4ms, 98.04%; 9.2MB, 7.14%
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int largestSumAfterKNegations(const vector<int> &A, int K) {
        multiset<int> tmp;
        int res = 0, _min = 0x7FFFFFFF;
        for (auto x : A) {
            res += x;
            if (x <= 0) tmp.insert(x);
            _min = min(_min, abs(x));
        }
        if (tmp.size() >= K) {
            auto it = tmp.begin();
            for (int i = 0; i < K; ++i) {
                res -= 2 * (*it);
                ++it;
            }
        } else {
            int x = K - tmp.size();
            auto it = tmp.begin();
            while (it != tmp.end()) {
                res -= 2 * (*it);
                ++it;
            }
            if (x & 1) res -= 2 * _min;
        }
        return res;
    }
};
